internal/resource/opener.go

```diff
@@ -284,20 +339,46 @@ func (ro ResourceOpener) set(chRes charmresource.Resource, reader io.ReadCloser)
 			_ = reader.Close()
 		}
 	}()
-	res, err := ro.state.SetResource(ro.appName, ro.retrievedBy.Id(), chRes, reader, false)
+
+	err = ro.resourceService.StoreResource(ctx, resource.StoreResourceArgs{
+		ResourceUUID:    resourceUUID,
+		Reader:          reader,
+		RetrievedBy:     ro.retrievedBy,
+		RetrievedByType: ro.retrievedByType,
+	})
 	if err != nil {
 		return resource.Resource{}, nil, errors.Capture(err)
 	}
 
 	// Make sure to use the potentially updated resource details.
-	res, reader, err = ro.open(res.Name)
+	res, reader, err := ro.resourceService.OpenResource(ctx, resourceUUID)
 	if err != nil {
 		return resource.Resource{}, nil, errors.Capture(err)
 	}
 
 	return res, reader, nil
 }
 
+// SetResource records that the resource is currently in use.
+func (ro ResourceOpener) SetResource(ctx context.Context, resName string) error {
```

Comment:
This isn't a great name. You could literally call it `SetResourceUsed`.

Reply:
Good point, thats better.